File: 1621-number-of-sets-of-k-non-overlapping-line-segments/1621-number-of-sets-of-k-non-overlapping-line-segments.ts
const MOD = 1000000007n;

function modPow(base: bigint, exp: bigint): bigint {
    let result = 1n;

    while (exp > 0n) {
        if (exp & 1n) {
            result = result * base % MOD;
        }

        base = base * base % MOD;

        exp >>= 1n;
    }

    return result;
}

function numberOfSets(n: number, k: number): number {
    const N = BigInt(n + k - 1);
    let R = BigInt(2 * k);

    if (R > N - R) {
        R = N - R;
    }

    let numerator = 1n;
    let denominator = 1n;

    for (let i = 1n; i <= R; i++) {
        numerator = numerator * (N - R + i) % MOD;

        denominator = denominator * i % MOD;
    }

    const inverseDenominator = modPow(denominator, MOD - 2n);

    return Number(numerator * inverseDenominator % MOD);
}

File: 1621-number-of-sets-of-k-non-overlapping-line-segments/1621-number-of-sets-of-k-non-overlapping-line-segments.go
package main

const MOD int64 = 1000000007

func modPow(base, exp int64) int64 {
	result := int64(1)

	for exp > 0 {
		if exp&1 == 1 {
			result = result * base % MOD
		}

		base = base * base % MOD

		exp >>= 1
	}

	return result
}

func numberOfSets(n int, k int) int {
	N := int64(n + k - 1)
	R := int64(2 * k)

	if R > N-R {
		R = N - R
	}

	numerator := int64(1)
	denominator := int64(1)

	for i := int64(1); i <= R; i++ {
		numerator = numerator * (N - R + i) % MOD

		denominator = denominator * i % MOD
	}

	inverseDenominator := modPow(denominator, MOD-2)

	return int(numerator * inverseDenominator % MOD)
}